pkg/placement/map-relation.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+// MapRelation2 is a relation represented by two map-based indices.
+// It is mutable.
+// It is not safe for concurrent access.
+type MapRelation2[First comparable, Second comparable] struct {
+	by1 map[First]MapSet[Second]
+	by2 map[Second]MapSet[First]
+}
+
+var _ MutableRelation2[string, float64] = &MapRelation2[string, float64]{}
+
+func NewMapRelation2[First comparable, Second comparable](pairs ...Pair[First, Second]) *MapRelation2[First, Second] {
+	return &MapRelation2[First, Second]{
+		by1: map[First]MapSet[Second]{},
+		by2: map[Second]MapSet[First]{},
+	}
+}
+
+func (mr MapRelation2[First, Second]) IsEmpty() bool {
+	return len(mr.by1) == 0
+}
+
+func (mr MapRelation2[First, Second]) LenIsCheap() bool { return false }
```

Comment:
Is it necessary (or worth it) to make `LenIsCheap` to be `true` ? I can imagine maintaining a counter to track the incremental change of length whenever a `MapRelation2` is updated, in place of `Relation2LenFromVisit`. But whether we need this might depends on how frequently `Len()` is called.

Reply:
Yeah, we can consider further development in the future.